### PR TITLE
Gemm + softmax (gemm + reduce_max + broadcast sub + exp + reduce_sum + broadcast div)

### PR DESCRIPTION
**Pull request by rocking5566**

### CODE REVIEW

**Comment by asroy** [comment]
This is only GEMM+softmax_reducion, need to add other kernels in this example too

**Comment by asroy** [comment]
This PR is doing GEMM and reduction separately.

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_1d.hpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                               const ElementwiseFunctor functor)
+    {
+        const auto a_global_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_a_global, a_grid_desc_m0.GetElementSpaceSize());
+        const auto b_global_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_b_global, b_grid_desc_m0.GetElementSpaceSize());
+        auto c_global_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_c_global, c_grid_desc_m0.GetElementSpaceSize());
+
+        StaticBuffer<AddressSpaceEnum::Vgpr, ADataType, ScalarPerVector, true> a_thread_buf;
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_1d.hpp:73]: Here the concept of `ThreadTileSize` and/or `ScalePerVector` is kind of vague.  Usually, `ThreadSliceSize` is used to indicate the number data to transfer by single `ThreadwiseTensorSliceTransfer` call.   `ScalePerVector` should be different from `ThreadSliceSize` since `ScalePerVector` is required to be on the vector dimension and should completely dividing the length of the vector dimension, but `ThreadSliceSize` can be different. 

Here I suggest: 

1.  Use `ThreadSliceSize` by `ThreadwiseTensorTransfer` and as the number of allocated vgprs
2.  If each thread has more than one slices to process, use some input parameter like `BlockChunkSize` to calculate the num_iter,   like  `num_iter = BlockChunkSize / (ThreadPerBlock * ThreadSliceSize)`.   But you don't have to support more than one slices per thread, since you can always allocate more workgroups.   In your case, `ThreadTileSize` is determined at compile-time,  so it seems you want each thread to handle one Tile/Slice to make the implementation easy

**Comment by qianfengz** [comment]
Also,  the file name `gridwise_elementwise_1d.hpp` should be changed to indicate the elementwise operator used here is `binary` operator, cause we might also use `unary` or `ternary` operate on the 1d tensor.  Names could be like

1. gridwise_1d_binary_operate.hpp
2. gridwise_1d_unary_operate.hpp
3. gridwise_1d_ternary_operate.hpp

My PR.182 and PR.192 also have similar elementwise binary/unary kernels defined.   By I think your implementation is more generic since your implementation support `ThreadSliceSize/ThreadTileSize`, but my implementation assumes each thread process only one element,  for int8 and fp16, use `ThreadSliceSize=4/2` could be beneficial for data access performance. 

After merging your P.R,  I will change my Batch-Norm forward codes to use your kernel

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_1d.hpp (hunk 1 above)
Comment (rocking5566) [reply include/ck/tensor_operation/gpu/grid/gridwise_elementwise_1d.hpp:73]: This is the thread mapping of implementation of elementwise kerenl
eltwise1d

In fact, I borrow the term of threadTile from tensile, which means the number of C element need to be calculated for each thread and this is so called tile size. 
Larger tile size means more works per thread and less work group.
More workgroups also consume the computational resource  
In this implementation, larger the `ThreadTileSize`, larger the iteration here.

`scalrPerVector `is the vector width of buffer_load_dword (buffer_load_dwordx2, buffer_load_dwordx4...etc)
I call `buffer_load_dwordxscalrPerVector`  in each iteration. 
Therefore,  the `SliceLength `and `SrcScalarPerVector `in `ThreadwiseTensorSliceTransfer_v2 `will be the same here.

**Comment by rocking5566** [comment]
> Also, the file name `gridwise_elementwise_1d.hpp` should be changed to indicate the elementwise operator used here is `binary` operator, cause we might also use `unary` or `ternary` operate on the 1d tensor. Names could be like
> 
> 1. gridwise_1d_binary_operate.hpp
> 2. gridwise_1d_unary_operate.hpp
> 3. gridwise_1d_ternary_operate.hpp
> 
> My PR.182 and PR.192 also have similar elementwise binary/unary kernels defined. By I think your implementation is more generic since your implementation support `ThreadSliceSize/ThreadTileSize`, but my implementation assumes each thread process only one element, for int8 and fp16, use `ThreadSliceSize=4/2` could be beneficial for data access performance.
> 
> After merging your P.R, I will change my Batch-Norm forward codes to use your kernel



> Also, the file name `gridwise_elementwise_1d.hpp` should be changed to indicate the elementwise operator used here is `binary` operator, cause we might also use `unary` or `ternary` operate on the 1d tensor. Names could be like
> 
> 1. gridwise_1d_binary_operate.hpp
> 2. gridwise_1d_unary_operate.hpp
> 3. gridwise_1d_ternary_operate.hpp
> 
> My PR.182 and PR.192 also have similar elementwise binary/unary kernels defined. By I think your implementation is more generic since your implementation support `ThreadSliceSize/ThreadTileSize`, but my implementation assumes each thread process only one element, for int8 and fp16, use `ThreadSliceSize=4/2` could be beneficial for data access performance.
> 
> After merging your P.R, I will change my Batch-Norm forward codes to use your kernel

Your suggestion is good, I will rename to gridwise_binary_elementwise_1d.hpp
Because elementwise operation is a term in deep learning area. 
This means there must be some sort of operation on the tensor.
[https://caffe.berkeleyvision.org/tutorial/layers/eltwise.html](url)

There are also another binary, ternary operation in deep learning, ex: concatenation.

In addition, as discussed with @asroy before.
We will rename original elementwise in each kernel parameter to the functor to stand for the fusion operation.
https://github.com/ROCmSoftwarePlatform/composable_kernel/issues/179

**Comment by rocking5566** [comment]
As discussed with @qianfengz  , I make following revise
1. Compute datatype, similar to accDatatype, but call acc here is confused
2. Add padding in MakeDescriptor_M0
3. rename elementwise to binary elementwise for file name and class name
4. To avoid loss of precision, use f32 as the acc datatype for reduce sum

File: example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+                                                        256,
+                                                        4,
+                                                        64,
+                                                        1,
+                                                        1,
+                                                        0,
+                                                        1,
+                                                        1>;
+
+struct Sub_Exp
```
Comment (asroy) [inline example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp:152]: SubExp

File: example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+// CAUSION - host reduce_max will call numeric_limits<ck::half_t>::lowest()
+// However, numeric_limits<ck::half_t>::lowest() will return zero. So, used half_float::half instead
+using HostReduceDataType = half_float::half;
+
+using ALayout = ck::tensor_layout::gemm::RowMajor;
+using BLayout = ck::tensor_layout::gemm::ColumnMajor;
+using CLayout = ck::tensor_layout::gemm::RowMajor;
+
+// clang-format off
+using DeviceGemmInstance = ck::tensor_operation::device::DeviceGemmXdl_C_Shuffle<
```
Comment (asroy) [inline example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp:54]: DeviceGemmXdl_C_Shuffle will be deprecated, please use DeviceGemm_Xdl_CShuffle

File: include/ck/tensor_operation/gpu/grid/gridwise_binary_elementwise_1d.hpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+          typename BDataType,
+          typename CDataType,
+          typename ComputeDataType,
+          typename GridDesc_M0,
+          typename ElementwiseFunctor,
+          index_t ScalarPerVector>
+struct GridwiseBinaryElementwise_1D
+{
+    static constexpr auto I0 = Number<0>{};
+    static constexpr auto thread_desc_M0 =
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/grid/gridwise_binary_elementwise_1d.hpp:43]: thread_desc_m0

File: include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+namespace ck {
+namespace tensor_operation {
+namespace device {
+
+template <typename ADataType,
+          typename BDataType,
+          typename CDataType,
+          typename ComputeDataType,
+          typename ElementwiseFunctor,
+          index_t ThreadPerBlock,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:18]: This should not be a template parameter

File: include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+          index_t ScalarPerVector>
+struct DeviceBinaryElementwise_2D : public DeviceBinaryElementwise<ElementwiseFunctor>
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:20]: DeviceBinaryElementwise_ND

You could make this Device Operation supporting N-D tensor (N=1~5).

File: include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp (hunk 5 above)
Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:18]: still not fixed

Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:18]: still not fixed

File: include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp (hunk 6 above)
Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:20]: Please open a JIRA task ticket and a github issue, and refer to this comment in both tickets.

Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:20]: Please open a JIRA task ticket and a github issue, and refer to this comment in both tickets.

File: include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp (hunk 5 above)
Comment (rocking5566) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp]: fix

File: include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp (hunk 6 above)
Comment (rocking5566) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:19]: https://ontrack-internal.amd.com/browse/LWPCK-190

Comment (rocking5566) [reply include/ck/tensor_operation/gpu/device/device_binary_elementwise_2d.hpp:19]: https://github.com/ROCmSoftwarePlatform/composable_kernel/issues/203

**Comment by asroy** [comment]
After PR https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/209 get merged, please fix issues in this PR (instead of in future PR)
1. refactor this PR to use fused GEMM+reduction
2. refactor device element-wise operation to support 1D~5D (https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/178#discussion_r855771367)

File: example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+        throw std::runtime_error(
+            "wrong! device_gemm with the specified compilation parameters does "
+            "not support this GEMM problem");
+    }
+
+    gemm_invoker.Run(gemm_argument, nrepeat);
+
+    // do reduce max
+    auto reduce_max                 = DeviceReduceMaxInstance{};
+    auto reduce_max_workaspace_size = reduce_max.GetWorkspaceSizeInBytes(c_m_n_shape, reduceDims);
```
Comment (aosewski) [inline example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp:376]:
```suggestion
    auto reduce_max_workspace_size = reduce_max.GetWorkspaceSizeInBytes(c_m_n_shape, reduceDims);
```

Comment (rocking5566) [reply example/19_gemm_softmax/gemm_softmax_xdl_fp16.cpp]: Good catch!
Fix in b6fe118ed0355b91956e6ad03ede09b55dc93264

**Comment by asroy** [comment]
closing this PR. new PR will be creatted
